src/mbgl/map/geometry_tile.cpp

```diff
@@ -0,0 +1,16 @@
+#include <mbgl/map/geometry_tile.hpp>
+#include <mbgl/style/filter_expression_private.hpp>
+
+namespace mbgl {
+
+mapbox::util::optional<Value> GeometryTileFeatureExtractor::getValue(const std::string& key) const {
+    if (key == "$type") {
+        return Value(uint64_t(feature.getType()));
+    }
+
+    return feature.getValue(key);
+}
+
+template bool evaluate(const FilterExpression&, const GeometryTileFeatureExtractor&);
```

Comment:
Could be improved — this was adapted from `master`: 

https://github.com/mapbox/mapbox-gl-native/blob/f82de98899a826bd3d5dc935c5a4e8a33c2d9b6b/src/mbgl/map/vector_tile.cpp#L176